PEP 692: Using TypedDict for more precise `**kwargs` typing

franekmagiera opened the pull request:
<!--

Please include the PEP number in the pull request title, example:

PEP NNN: Summary of the changes made

In addition, please sign the CLA.

For more information, please read our Contributing Guidelines (CONTRIBUTING.rst)

-->

The first version of the "Using TypedDict for more precise **kwargs typing" PEP

AA-Turner commented:
You can take 692.

A

AA-Turner commented on pep-9999.rst:
```diff
@@ -0,0 +1,344 @@
+PEP: 9999
+Title: Using TypedDict for more precise **kwargs typing
```
Escape asterisk
```suggestion
Title: Using TypedDict for more precise \*\*kwargs typing
```

AA-Turner reviewed (changes requested):
To fix CI:

A

franekmagiera commented:
Thank you @AA-Turner!

AA-Turner commented on .github/CODEOWNERS:
```diff
@@ -658,4 +659,4 @@ pep-8016.rst  @njsmith @dstufft
 pep-8100.rst  @njsmith
 # pep-8101.rst
-# pep-8102.rst
+# pep-8102.rst
```
You've deleted the trailing newline here.

AA-Turner commented:
You will also need to sign the CLA -- you've used your Goldman email in commits, you may need to check if you need to use the corporate CLA vs or if you can use the individual CLA.

A

AA-Turner commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+Function calls with standard dictionaries
+-----------------------------------------
+
+Calling a function that has ``**kwargs`` typed using the ``**kwargs: **Movie`` sytnax with a standard dictionary must raise an error. For example:
```
Must raise an error at runtime (the python level) or when type checking? (The former doesn't make sense, but it is unspecified currently)

erictraut replied:
Yes, I presume you mean "a type checker should generate an error". The term `raise` implies a runtime error.

AA-Turner commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+
+    def foo(**kwargs: **Movie) -> None: ...
+
+would mean that the ``**kwargs`` comprise two keyword arguments specified by ``Movie`` (i.e. a ``name`` keyword of type ``str`` and a ``year`` keyword of type ``int``). Then, inside the function itself, appropriate type checking would take place:
```
"appropriate type checking would take place" at type checking time, not at runtime.

AA-Turner commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+Keyword collisions
+------------------
+
+A ``TypedDict`` that is used to type ``**kwargs`` could potentially contain keys that are already defined in the function's signature. If the duplicate name is a standard argument, an error should be raised. If the duplicate name is a positional only argument, no errors should be raised. For example:
```
"an error should be raised" / "no errors should be raised" -- again, unspecified

JelleZijlstra replied:
All of this should be type checking-only. Maybe just put a note at the beginning of the spec saying that the new annotation has no runtime effect.

erictraut replied:
As mentioned above, I'd avoid using the term "raise" when talking about errors emitted by a type checker. Use "generated" or "emitted" or "reported" instead. The term "raise" implies a runtime exception.

AA-Turner commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+        title: str
+        year: NotRequired[int]
+
+When using a ``TypedDict`` to type ``**kwargs`` all of the required and non-required keys should correspond to required and non-required function keyword parameters. Therefore, if a required key is not supported by the caller, then an error must be raised.
```
"an error must be raised" -- by whom?

erictraut replied:
Same as above — avoid the term "raise".

AA-Turner commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+Required and non-required keys
+------------------------------
+
+By default all keys in a ``TypedDict`` are required. This behaviour can be overriden by setting the dictionary's ``total`` parameter as ``False``:
+
+::
+
+    class Movie(TypedDict, total=False):
+        title: str
+        year: int 
+
+    dictionary: Movie = {"title": "Life of Brian"}         # OK! "year" is not required.
+    dictionary = {"title": "Life of Brian", "year": 1979}  # OK!
+
+Moreover, PEP 655 introduced new type qualifiers - ``typing.Required`` and ``typing.NotRequired`` - that enable specyfing whether a particular key is required or not:
```
I would combine these two paragraphs and remove the example -- in the Specification section, a level of familiarity with the type hinting system can be assumed.

AA-Turner commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+
+Assignment
+----------
+A function typed using the ``**kwargs: **Movie`` construct can be assigned to another callable type only if they are compatible. This can happen in the following cases:
```
"assigned to another callable type" -- I think this could be clearer. You're talking about valid assignments under static type checking, not runtime assignments. Perhaps "validly assigned" or "pass type checking" or similar.

AA-Turner commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+    def foo(**kwargs: **TypedDictUnion) -> None: ...  # WRONG! Cannot use a union of TypedDicts to type **kwargs
+
+
+Additional keys
+---------------
+
+A big advantage of using ``**kwargs`` is the ability to pass arbitrary dictionaries containing keyword-value pairs to the function and let the function pick and use only the ones it needs. However, currently, there is no way to define a ``TypedDict`` that would allow additional keys, so supporting this feature would require an addition to the ``TypedDict`` specification. If this limitation blocks ...
```
This section seems speculative, I would remove personally, or add as a limitation / rejected idea. It shouldn't be in Specification.

JelleZijlstra replied:
Agree

franekmagiera commented:
I need to be using a corporate CLA  that AFAIK has been signed - however I'll need to verify if there is still something for me to do, still new to our internal process.

AA-Turner reviewed (changes requested):
I think you need to make the distinction between runtime and static type checking much clearer -- it is usually clear what you mean, but I have to work somewhat to get to your intent.

I'd strongly reccomend adding a rejected ideas section, as well as how to teach this, especially as this is primarily a user facing change.

Your rationale section is very thin -- a simple counterargument would be "why not use `dataclasses`" -- you should address why specifically using `kwargs` with `TypedDict` is a useful addition to the language, especially given it connotes grammar changes.

AA-Turner commented:
Also cc @AlexWaygood for interest -- I've seen you doing various typing things but you might not be subscribed to the peps repo.

A

AlexWaygood commented:
> Also cc @AlexWaygood for interest -- I've seen you doing various typing things but you might not be subscribed to the peps repo.
> 
> A

Too slow, I've already read it ;)

I'm excited to see movement in this area! Will try to give some comments in the next few days, either here and/or on typing-sig.

JelleZijlstra commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+Motivation
+==========
+
+Currently annotating ``**kwargs`` with a type ``T`` means that the ``kwargs`` type is in fact ``Dict[str, T]``. For example:
```
```suggestion
Currently annotating ``**kwargs`` with a type ``T`` means that the ``kwargs`` type is in fact ``dict[str, T]``. For example:
```

JelleZijlstra commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+
+    def foo(**kwargs: str) -> None: ...
+
+means that all keyword arguments in ``foo`` are strings (i.e. ``kwargs`` is of type ``Dict[str, str]``). This behaviour limits the ability to type annotate ``**kwargs`` only to the cases where all of them are of the same type. However, it is often the case that keyword arguments conveyed by ``**kwargs`` have different types that are dependent on the keyword's name. In those cases type annotating  ...
```
```suggestion
means that all keyword arguments in ``foo`` are strings (i.e., ``kwargs`` is of type ``dict[str, str]``). This behaviour limits the ability to type annotate ``**kwargs`` only to the cases where all of them are of the same type. However, it is often the case that keyword arguments conveyed by ``**kwargs`` have different types that are dependent on the keyword's name. In those cases type annotating ``**kwargs`` is not possible. This is especially a problem for already existing codebases where the need of refactoring the code in order to introduce proper type annotations may be considered not worth the effort. This in turn prevents the project from getting all of the benefits that type hinting can provide. As a consequence, there has been a lot of discussion around supporting more precise ``**kwargs`` typing [#mypyIssue4441]_ and it became a feature that is necessary for a large part of Python community.
```

JelleZijlstra commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+    
+    def foo(**kwargs: Movie) -> None: ...
+
+means that each keyword argument in ``foo`` is itself a ``Movie`` dictionary that has a ``name`` key with a string type value and a ``year`` key with an integer type value. Therefore, in order to support specifing ``kwargs`` type as a ``TypedDict`` without breaking current behaviour, a new syntax has to be introduced.
```
```suggestion
means that each keyword argument in ``foo`` is itself a ``Movie`` dictionary that has a ``name`` key with a string type value and a ``year`` key with an integer type value. Therefore, in order to support specifing ``kwargs`` type as a ``TypedDict`` without breaking current behaviour, a new syntax has to be introduced.
```
```suggestion
means that each keyword argument in ``foo`` is itself a ``Movie`` dictionary that has a ``name`` key with a string type value and a ``year`` key with an integer type value. Therefore, in order to support specifying ``kwargs`` type as a ``TypedDict`` without breaking current behaviour, a new syntax has to be introduced.
```

JelleZijlstra commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+::
+
+    def foo(**kwargs: **Movie) -> None:
+        kwargs["name"].capitalize()      # OK!
```
Could use `assert_type(kwargs, Movie)` to communicate the point more succinctly.

JelleZijlstra commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+Function calls with standard dictionaries
+-----------------------------------------
+
+Calling a function that has ``**kwargs`` typed using the ``**kwargs: **Movie`` sytnax with a standard dictionary must raise an error. For example:
```
```suggestion
Calling a function that has ``**kwargs`` typed using the ``**kwargs: **Movie`` syntax with a standard dictionary must raise an error. For example:
```

JelleZijlstra commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+
+    def foo(**kwargs: **Movie) -> None: ...
+
+    movie: dict[str, str] = {"name": "Life of Brian", "year": 1979}
```
This line should also raise an error, because 1979 is not a str.
```suggestion
    movie: dict[str, object] = {"name": "Life of Brian", "year": 1979}
```

JelleZijlstra commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+    foo(**typedMovie)  # OK!
+
+
+However, if the dictionary is completely untyped it is up to the type checker whether to raise an error or not:
```
I think we can simply omit this example and paragraph.

erictraut replied:
I agree that this should be omitted. This gets into inference behaviors, which are not specified in any other PEPs.

JelleZijlstra commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+Required and non-required keys
+------------------------------
+
+By default all keys in a ``TypedDict`` are required. This behaviour can be overriden by setting the dictionary's ``total`` parameter as ``False``:
```
```suggestion
By default all keys in a ``TypedDict`` are required. This behaviour can be overridden by setting the dictionary's ``total`` parameter as ``False``:
```

JelleZijlstra commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+----------
+A function typed using the ``**kwargs: **Movie`` construct can be assigned to another callable type only if they are compatible. This can happen in the following cases:
+
+1. Both destination and source have a ``**kwargs: **TypedDict`` parameter, the destination ``TypedDict`` is assignable to the source ``TypedDict`` and the rest of the parameters are the compatible.
```
```suggestion
1. Both destination and source have a ``**kwargs: **TypedDict`` parameter, the destination ``TypedDict`` is assignable to the source ``TypedDict`` and the rest of the parameters are compatible.
```

JelleZijlstra replied:
Also, I think this gets the variance wrong. Consider two callables `f1: (int) -> bool` and `f2: (object) -> bool`. `f1` is *not* assignable to `f2`, because when you have an `f2`, you can pass it any object, but `f1` must have an int. On the other hand, `f2` is assignable to `f1`. Functions parameters behave contravariantly.

For analogous reasons, for two functions `(**TD1) -> ...` and `(**TD2) -> ...`, the first function is assignable to the second if `TD2` is assignable to `TD1`, not the other way around.

(2) also gets the variance wrong.

Last, you ignore the return type, but it also affects assignability. However, it is covariant.

franekmagiera replied:
I think the way I worded it was confusing.

> For analogous reasons, for two functions `(**TD1) -> ...` and `(**TD2) -> ...`, the first function is assignable to the second if `TD2` is assignable to `TD1`, not the other way around.

The first function is assignable to the second, so the first function is the "source" and the second one is the "destination". So the destination `TypedDict` (`TD2`) has to be assignable to the source TypedDict (`TD1`) - I think this is the source (pun not intended) of the error - what I meant is:

`Both destination and source have a ``**kwargs: **TypedDict`` parameter, the destination function's ``TypedDict`` is assignable to the source function's ``TypedDict`` and the rest of the parameters are compatible.`

Let me try to find a better way to phrase it.

JelleZijlstra commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+
+5. If the destination callable contains ``**kwargs: **TypedDict`` then source callable should have keyword parameters assignable to the key value pairs in the ``TypedDict``. Again, the rest of the parameters should be compatible.
+
+``TypedDict`` unions
```
This should go in "Rejected Ideas"

JelleZijlstra commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+    >>> foo.__annotations__
+    {'kwargs': **Movie}
+
+In order to make the runtime behaviour consistent with the AST as shown above we propose to:
```
We shouldn't specify what goes in typing.py: that's an implementation detail. But we *should* specify what the `**` operator does, which is call the `__unpack__` special method. We should make it explicit that `def f(**kwargs: **T):` is equivalent to `def f(**kwargs: T.__unpack__()):`.

JelleZijlstra commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+            return self._unpacked
+
+
+Implications
```
Not sure we even need this section.

JelleZijlstra reviewed (commented):
Thanks! I think the PEP text still needs some work. I have a few suggestions, but I'll probably need another pass.

Also, please make your lines shorter by breaking them to around 70-80 characters. That will make editing easier.

erictraut commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+Specification
+=============
+
+To support the aforementioned use case we propose to use the double asterisk syntax inside of the type annotation. The required grammar change is discussed in more detail in section `Grammar Changes`_. Continuing the previous example:
```
I didn't realize that you were thinking of proposing a grammar change. The SC has indicated that the bar is _very_ high when it comes to grammar changes for features that are specific to typing. I therefore think it's likely that this proposal will be rejected if it proposes a grammar change. I thought the proposal was to support `Unpack[T]`, which doesn't require any new syntax.

AlexWaygood replied:
I think it would be good to engage with the Steering Council on this prior to submitting the PEP, since they did, after all, accept PEP 646 (also required syntax changes, also very typing-specific). The changes in syntax here feel like a fairly natural extension of the changes made by PEP 646, and don't feel nearly as jarring as those proposed by PEP 677 (a PEP I initially supported, but had mixed feelings about by the end).

JelleZijlstra replied:
Yes, I'm hopeful this will be acceptable because PEP 646's similar change seemed uncontroversial. It's true though that the SC at the PyCon panel recommended to split out syntax changes from typing ecosystem changes, so we could consider having one PEP for the new type system feature (initially spelled with `Unpack`) and another for the grammar change. Let me think about this some more and ask for other opinions.

erictraut commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+
+1. Both destination and source have a ``**kwargs: **TypedDict`` parameter, the destination ``TypedDict`` is assignable to the source ``TypedDict`` and the rest of the parameters are the compatible.
+
+2. If the destination callable contains traditionally typed ``**kwargs: T`` and the source callable is typed using ``**kwargs: **TypedDict`` then each ``TypedDict`` value has to be assignable to type ``T``. Again, the rest of the parameters has to be compatible.
```
This section also gets the variance backward.

erictraut commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+
+    bar = foo  # OK!
+
+4. If the destination callable contains ``**kwargs: **TypedDict`` and the source callable containes either untyped or traditionally typed ``**kwargs: T`` then an error should be raised.
```
Avoid the term "raised" here.

erictraut commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+
+2. If the destination callable contains traditionally typed ``**kwargs: T`` and the source callable is typed using ``**kwargs: **TypedDict`` then each ``TypedDict`` value has to be assignable to type ``T``. Again, the rest of the parameters has to be compatible.
+
+3. If the destination callable doesn't contain ``**kwargs`` and the source callable contains ``**kwargs: **TypedDict`` then all of the ``TypedDict`` fields should be assignable to individual named parameters. Again, the rest of the parameters has to be compatible. For example:
```
Minor nit: "has to be" should be "have to be".

erictraut commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+Alternatives
+------------
+
+Instead of making the grammar change, ``Unpack`` could be the only way to annotate ``**kwargs`` of different types. However, introducing the double asterisk syntax has two advantages. Namely, it is more concise and more intuitive than using ``Unpack``.
```
The "more concise and more intuitive" argument is unlikely to be enough to justify a grammar change here. This is an advanced typing feature that will be used by very few Python programmers, so making it slightly more concise and intuitive is not a strong argument.

AlexWaygood replied:
For reference, here is PEP 646's justification for its proposed grammar changes (PEP 646 was accepted by the Steering Council): https://peps.python.org/pep-0646/#alternatives-why-not-just-use-unpack

franekmagiera commented:
Thank you for feedback and all the comments! I'll try to address them by Thursday/Friday.

franekmagiera commented:
Thanks for all the comments once again, they are very helpful! I addressed a part of them, will push the changes once I have more.

davidfstr commented:
> I'm excited to see movement in this area!

I'll also echo the sentiment that I'm excited to see a concrete proposal for [TypedDict for **kwargs typing](https://github.com/python/mypy/issues/4441). 😁

I'll plan to take a look at the next draft as well.

Bluenix2 reviewed (commented):
Hello, it appears that I forgot to press the "Submit review" button

davidfstr commented on pep-0692.rst:
```diff
@@ -0,0 +1,438 @@
+    movie: dict[str, object] = {"name": "Life of Brian", "year": 1979}
+    foo(**movie)  # WRONG! Movie is of type dict[str, object]
+
+    typedMovie: Movie = {"name": "The Meaning of Life", "year": 1983}
+    foo(**typedMovie)  # OK!
```
```suggestion
    typed_movie: Movie = {"name": "The Meaning of Life", "year": 1983}
    foo(**typed_movie)  # OK!
```
Nit: [Camelcase is not for Python.](https://www.youtube.com/watch?v=hgI0p1zf31k)

davidfstr commented on pep-0692.rst:
```diff
@@ -0,0 +1,438 @@
+
+A ``TypedDict`` that is used to type ``**kwargs`` could potentially contain keys
+that are already defined in the function's signature. If the duplicate name is a
+standard argument, an error should be reported. If the duplicate name is a
```
```suggestion
standard argument, an error should be reported by type checkers. If the duplicate name is a
```

davidfstr commented on pep-0692.rst:
```diff
@@ -0,0 +1,438 @@
+
+By default all keys in a ``TypedDict`` are required. This behaviour can be
+overridden by setting the dictionary's ``total`` parameter as ``False``.
+Moreover, PEP 655 introduced new type qualifiers - ``typing.Required`` and
```
```suggestion
Moreover, :pep:`655` introduced new type qualifiers - ``typing.Required`` and
```
Suggest making the PEP link explicit. (I can't remember if regular text in format "PEP ###" is auto-linked or not.)

davidfstr commented on pep-0692.rst:
```diff
@@ -0,0 +1,438 @@
+When using a ``TypedDict`` to type ``**kwargs`` all of the required and
+non-required keys should correspond to required and non-required function
+keyword parameters. Therefore, if a required key is not supported by the caller,
+then an error must be reported.
```
```suggestion
then an error must be reported by type checkers.
```

davidfstr commented on pep-0692.rst:
```diff
@@ -0,0 +1,438 @@
+``**kwargs: **TypedDict`` then an error should be generated, because
+traditionally typed ``**kwargs`` aren't checked for keyword names.
+
+To sum up points, functions' parameters should behave contravariantly. In
+addition, functions' return types should behave covariantly. 
```
```suggestion
To summarize, function parameters should behave contravariantly and
function return types should behave covariantly. 
```
Nit: Wordsmithing

davidfstr commented on pep-0692.rst:
```diff
@@ -0,0 +1,438 @@
+enable more precise ``**kwargs`` typing. The new approach revolves around using
+``TypedDict`` to type ``**kwargs`` that comprise keyword arguments of different
+types.
+
```
Might be worth highlighting near the beginning that:

* a Python grammar change is being proposed

Since that is useful context for a reader to determine whether they need to respond to changes in the PEP.

adam-grant-hendry replied:
It would be nice if a note could be added that

> Usage of Unpack for *args and TypeVarTuple is proposed in PEP 589.

or an update note added to `PEP 589` to the effect of:

> **UPDATE:** Usage of Unpack also for **kwargs and TypedDict is proposed in PEP 692.

to link the two and provide greater traceability. It's a little unclear to me whether or not `Unpack` will work for all 4.

franekmagiera replied:
`Unpack` isn't mentioned until the Backwards Compatibility section where it is linked to PEP 646. Added a note that the new use cases of `Unpack` should not interfere with the behaviour described in 646.

davidfstr commented on pep-0692.rst:
```diff
@@ -0,0 +1,438 @@
+        kw_defaults=[],
+        kwarg=arg(
+          arg='kwargs',
+          annotation=DoubleStarred(
+            value=Name(id='Movie', ctx=Load()),
+            ctx=Load())),
```
So a `DoubleStarred(...)` would appear here if you wrote `def foo(**kwargs: **Movie): ...`.

What would appear instead if you wrote `def foo(**kwargs: Unpack[Movie]): ...`? (This may be a useful example to distinguish how both forms would appear in the AST, if they were to appear differently.)

franekmagiera replied:
Added the AST with Unpack in the Backwards Compatibility section.

davidfstr commented on pep-0692.rst:
```diff
@@ -0,0 +1,438 @@
+    >>> foo.__annotations__
+    {'kwargs': **Movie}
+
+The double asterisk operator should call the ``__unpack__`` special method on
+the object it was used on. This means that ``def foo(**kwargs: **T): ...`` is
+equivalent to ``def foo(**kwargs: T.__unpack__()): ...``.
```
This appears to imply that an entirely new `**expr` operator (with related `__unpack__` dunder) is being introduced, that could potentially be used _anywhere_. I would advocate narrowing the scope of the PEP to only allow the new `**expr` syntax in the position of a kwarg annotation and not (yet) elsewhere else.

davidfstr replied:
Based on reading other comments it sounds like there is _not_ an intent to introduce a new general operator. In that case I might suggest replacing the word "operator" here with some other word like "syntax":

```suggestion
The double asterisk syntax should call the ``__unpack__`` special method on
the object it was used on. This means that ``def foo(**kwargs: **T): ...`` is
equivalent to ``def foo(**kwargs: T.__unpack__()): ...``.
```

davidfstr commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+    >>> def foo(**kwargs: **Movie): ...
+    ...
+    >>> foo.__annotations__
+    {'kwargs': **Movie}
```
I might suggest having the example output be instead:
```
{'kwargs': Unpack[Movie]}
```
which avoids introducing the new expression form `**expr`.

davidfstr reviewed (changes requested):
Great first draft! 

My most significant feedback is around a resistance to introducing a new `**expr` operator and related `__unpack__` dunder method. I think this PEP would have a better chance of passing core devs and the SC with a more narrow proposal that _only_ accepts the new `**expr` syntax in the position of a kwargs type annotation.

franekmagiera commented on pep-0692.rst:
```diff
@@ -0,0 +1,344 @@
+
+    def foo(**kwargs: Unpack[Movie]) -> None: ...
+
+There are several reasons for reusing PEP 646's ``Unpack``. Firstly, the name is quite suitable and intuitive for the ``**kwargs`` typing use case as the keywords arguments are "unpacked" from the ``TypedDict``. Secondly, there would be no need to introduce any new special forms. Lastly, the use of ``Unpack`` for the purposes described in this PEP does not interfere with the use cases described i ...
```
I don't think there is a need for that - in the case of `Unpack` the AST should be consistent with the runtime annotations. The argument for doing that would be to keep `Unpack[Movie]` and `**kwargs: **Movie` consistent, but then `Unpack` would probably have to call that as well when being used in different contexts - it feels like it might be worth it to keep it as so that `Unpack` had the flexibility for other use cases?

franekmagiera commented:
Big thanks for all the comments! Sorry for the delayed answer, I've been on vacation with limited computer/internet access, a lot of driving etc. -- I've updated as much as I could, will try to pick up things that I missed tomorrow.

Most significant changes:
 - Made the changes to avoid the term "raise" and tried to mention that any "errors" should be understood as "type checker errors"
 - Refactored the "Assignments" section

Regarding the `__unpack__` method - my understanding/intention was as follows:
1. `**kwargs: **Movie` should generate a new `DoubleStarred` AST node so that type checkers can discern between the "old" and new/proposed `**kwargs` type hinting
2. For the benefit of tools that use annotations at runtime `__annotations__` should be consistent with the AST
3. So the `**Movie` in `**kwargs: **Movie` context should be converted to an object that would enable the runtime tools to know that the type hint was used for more precise **kwargs typing - that could be either `Unpack[Movie]` or `_UnpackedTypedDict` - not sure if the PEP should be specyfing that - and `repr` of that object should prepend `**` to the name of the TypedDict that was used.
4. The idea was to make the `DoubleStarred` AST node generate code that would call `__unpack__` on e.g. `Movie` and this would return an object described in point 3

franekmagiera commented:
Thanks again for all the comments here and on the typing sig mailing list! I tried to incorporate all of the suggestions and think the draft should be ready for another look. The most significant changes revolve around the assignment section, a new section on Passing kwargs inside the function to another function and using final `TypedDict`s in certain contexts.

adam-grant-hendry commented on pep-0692.rst:
```diff
@@ -0,0 +1,599 @@
+However, it is often the case that keyword arguments conveyed by ``**kwargs``
+have different types that are dependent on the keyword's name. In those cases
+type annotating ``**kwargs`` is not possible. This is especially a problem for
+already existing codebases where the need of refactoring the code in order to
```
Based on the conversation from “ Support for default types in TypedDict #1165”, I’m disinclined to believe the additional type hinting provided would be of benefit. The kwargs are hidden (abstracted) within a TypedDict, so the type hinting provided (esp. by certain language servers) doesn't provide useful information to the caller on what arguments to pass.

I recommend a note be added that this PEP is born out of convenience, not necessity, and that it be strongly encouraged wherever possible that the full argument list be type hinted.

franekmagiera replied:
Thanks for bringing that up!

>I’m disinclined to believe the additional type hinting provided would be of benefit.

In my opinion the proposal could bring benefits, although I agree in some cases (new codebases where the keyword arguments are known from the start/typing via stubs) using "explicit" signatures is a better tool for the job.

>The kwargs are hidden (abstracted) within a TypedDict, so the type hinting provided (esp. by certain language servers) doesn't provide useful information to the caller on what arguments to pass.

Proposed type hinting would show the caller exactly what kind of keyword arguments the function expects. It is just the case of looking up how the TypedDict is defined (not an expert on language servers but seems like those could deal with "unpacking" the kwargs typed this way and provide a more user friendly hint). As I've mentioned in the response to the thread you linked, realistically not everyone will want to refactor and in that case I'd much rather see `**kwargs: **SomeTypedDictIHaveToLookUp` than `**kwargs: Any`. Also then, `**kwargs` could be treated as a `TypedDict` and type checked inside the function itself without the need of refactoring.

I think the discussion uncovered some valuable points that are worth adding to the draft (when it is better to use something else than what is being proposed). Let me try to include that.

a-reich commented on pep-0692.rst:
```diff
@@ -198,19 +211,52 @@ Continuing the previous example:
+    dest = src      # WRONG!
+    dest(**animal)  # Fails at runtime.
+
+However, when the ``TypedDict`` in the function signature is final, we are
```
Question: The idea that decorating a TD class with @final ensures its values will not have additional keys doesn’t make sense to me. PEP 589 says that TD compatibility is based on structural, not nominal, subtyping. So whether another TD *subclasses* the intended one (as prevented by final) is irrelevant. 

Here’s an [example](https://mypy-play.net/?mypy=latest&python=3.10&gist=084cc24abbae44aa982bd2255cb325d5).

franekmagiera replied:
Right, I was thinking about that, thanks for mentioning that. At the time it seemed like a good idea because I thought that as a user you would have to "hack around" to break that intentionally, but probably that's not a good premise to base the specification on. On second thought it seems like it would make more sense to disallow that. Let me make the changes.

davidfstr commented on pep-0692.rst:
```diff
@@ -0,0 +1,643 @@
+By default all keys in a ``TypedDict`` are required. This behaviour can be
+overridden by setting the dictionary's ``total`` parameter as ``False``.
+Moreover, :pep:`655` introduced new type qualifiers - ``typing.Required`` and
+``typing.NotRequired`` - that enable specyfing whether a particular key is
```
```suggestion
``typing.NotRequired`` - that enable specifing whether a particular key is
```

Nit: Spelling

davidfstr commented on pep-0692.rst:
```diff
@@ -0,0 +1,643 @@
+contract via type hints. 
+
+Adding type hints directly in the source code as opposed to the ``*.pyi``
+stubs benefits to anyone who reads the code as it is easier to understand.
```
```suggestion
stubs benefits anyone who reads the code as it is easier to understand.
```
Nit: Wordsmith

davidfstr commented on pep-0692.rst:
```diff
@@ -0,0 +1,643 @@
+
+    double_star_expression: '**' expression
+
+A new AST node needs to be created so that type checkers can differentiate the
+semantics of the new syntax and take into account that the ``**kwargs`` should
```
> differentiate the semantics of the new syntax

I didn't understand what this fragment was trying to say. Differentiate what from what else?

davidfstr reviewed (approved):
Looking good. 👍  Just left a few remaining comments.

franekmagiera commented:
@davidfstr Thanks for the review! Added suggested changes.

CAM-Gerlach commented:
FYI, @franekmagiera — you can apply a single suggested change by simply clicking `Commit` on the suggested change, and apply them all in one commit (usually better) by going to the `Files` tab, clicking `Add to batch` on each change and then `Commit` at the top. This automatically commits the change the reviewer requested, avoiding any mistakes, and auto-resolves the requested change, making things easier for both you and the reviewers.

CAM-Gerlach reviewed (commented):
Hey @franekmagiera , thanks for submitting this PEP! It looks like its getting close to ready. Here I have a smattering of minor changes, nearly all as GitHub suggestions, covering PEP formatting, reST syntax and a couple related things.

I did notice some grammar, typographical, clarity and phrasing issues, but I only opportunistically fixed unambiguous ones on the lines that I otherwise touched. If you'd like me to do a full copyedit, I'd be happy to; to avoid delaying and further cluttering this PR, I can make them in a separate PR as a followup to this one, which you can then review.

Note: You can apply changes automatically by clicking `Add to batch` on each suggestion in the `Files` tab, and then once all the ones you want are selected, clinking `Commit` at the top). This will ensure there are no unintentional mistakes when transposing the change, and will auto-resolve the associated review comments. Thanks!

CAM-Gerlach commented on pep-0692.rst:
Seems like this one might have gotten lost:

```suggestion
``TypedDict`` as a `pull request to mypy <mypyPull10576_>`__
and `to mypy_extensions <mypyExtensionsPull22_>`__.
The implementation uses ``Expand`` instead of ``Unpack``.
```

Make these links, well, links, instead of footnotes.

franekmagiera commented:
@CAM-Gerlach thanks for all the tips!

>If you'd like me to do a full copyedit, I'd be happy to; to avoid delaying and further cluttering this PR, I can make them in a separate PR as a followup to this one, which you can then review.

That would be great, all the suggestions really improved the quality of the PR, thank you!

franekmagiera commented on pep-0692.rst:
```diff
@@ -0,0 +1,645 @@
+This can happen in the following cases:
+
+1. Both destination and source functions have a ``**kwargs: **TypedDict``
+parameter and the destination function's ``TypedDict`` is assignable to the
```
Makes sense, will try to pick that up tomorrow (I'm on the road again with limited access to Internet and my computer).

franekmagiera replied:
In the end used subsections for that.

Bluenix2 commented:
Where can I find the discussion (if there's been one) behind the naming of the dunder? It occured to me today that while `__unpack__()` perfectly fits what we expect to happen, how will the rest of the Python ecosystem react to this dunder?

I mean, if you were to read `__unpack__()` defined somewhere without any knowledge of this PEP (but perhaps aware that the class supported `**` in the annotations) would you immediately connect that with `**kwargs: **SubclassTypedDict`?

---

I do not immediately have any suggestions apart from working around it by not adding a new dunder - because even using for example `__unpack_kwargs__()` could be confused with `func(**var)`.

One work-around though, is also using `__iter__()` since users are able to differentiate it from `*TypeVarTuple` in introspection because the argument will be a `VAR_KEYWORD` and not `VAR_POSITIONAL`. The downside would be that at runtime there would be no error doing `*args: *SubclassTypedDict` and type checkers would need to special-case the unpacking to only allow `*` on some classes and `**` on other classes.

AlexWaygood commented:
> Where can I find the discussion (if there's been one) behind the naming of the dunder? It occured to me today that while `__unpack__()` perfectly fits what we expect to happen, how will the rest of the Python ecosystem react to this dunder?

The language reference states that all dunder names are reserved for use by the core language and the stdlib; there are no backwards-compatibility guarantees with respect to dunder names. If third-party libraries or apps are using `__unpacked__` for anything, that's on them.

franekmagiera commented:
@Bluenix2 unfortunately there hasn't been any discussion around naming the dunder. The idea was that similarly to `~x` calling `x.__invert__()`, `**kwargs: **T` would call `T.__unpack__()`

JelleZijlstra commented on pep-0692.rst:
```diff
@@ -0,0 +1,594 @@
+``Unpack`` for the purposes described in this PEP does not interfere with the
+use cases described in :pep:`646`.
+
+It is worth pointing out that the AST generated for ``Unpack`` would differ
```
This seems unnecessary since this AST is standard for all Python 3 versions (Unpack isn't syntactically special) and we're not changing anything to it. I'd remove this paragraph.

franekmagiera replied:
Right, makes sense, removed.

JelleZijlstra reviewed (commented):
I think this PR has run its course and it's time to merge, but the CLA check is failing. @franekmagiera did you sign the CLA?

franekmagiera commented:
@JelleZijlstra the corporate CLA was signed, I was told that it is ok to sign for me the individual contributor license agreement to unblock the pipeline, so went ahead and did that. Also removed the section on AST for Unpack.

Bluenix2 commented:
> > The language reference states that all dunder names are reserved for use by the core language and the stdlib; there are no backwards-compatibility guarantees with respect to dunder names. If third-party libraries or apps are using `__unpack__` for anything, that's on them.
> 
> Reference for this: https://docs.python.org/3/reference/lexical_analysis.html#reserved-classes-of-identifiers. The docs state:
> 
> > `__*__`
> > System-defined names, informally known as “dunder” names. These names are defined by the interpreter and its implementation (including the standard library). Current system names are discussed in the [Special method names](https://docs.python.org/3/reference/datamodel.html#specialnames) section and elsewhere. More will likely be defined in future versions of Python. _Any_ use of `__*__` names, in any context, that does not follow explicitly documented use, is subject to breakage without warning.

This is besides my point - I am aware of this policy. @franekmagiera answered my question though, which was about the specific *name* of the dunder.

franekmagiera commented:
@CAM-Gerlach now that this is merged, could I take you up on your offer to do a full copyedit?

CAM-Gerlach commented:
> @CAM-Gerlach now that this is merged, could I take you up on your offer to do a full copyedit?

Sure, added to my queue. I have some other things to take care, but I'll try to get to it today or tomorrow. Thanks!

rafalkrupinski commented:
Why not using Protocol for this? It would allow extra arguments

CAM-Gerlach commented:
@rafalkrupinski you might want to bring that up on the [PEP's canonical discussion thread](https://mail.python.org/archives/list/typing-sig@python.org/thread/U42MJE6QZYWPVIFHJIGIT7OE52ZGIQV3/) instead
